server/app/tasks/DatabaseSeedTask.java

```
@@ -0,0 +1,162 @@
+      versionRepository.get().getDraftVersion();
+      inSerializableTransaction(this::seedCanonicalQuestions, 1);
+    }
+  }
+
+  /**
+   * Ensures that questions with names matching those in {@code CANONICAL_QUESTIONS} are present in
+   * the database, inserting the definitions in {@code CANONICAL_QUESTIONS} if any aren't found.
+   */
+  private void seedCanonicalQuestions() {
+    var questionService = questionServiceProvider.get();
+    ImmutableSet<String> existingQuestionNames = questionService.getQuestionNames();
+
+    for (QuestionDefinition questionDefinition : CANONICAL_QUESTIONS) {
+      if (existingQuestionNames.contains(questionDefinition.getName())) {
+        LOGGER.info(
+            "Canonical question \"%s\" exists at server start", questionDefinition.getName());
+        continue;
+      }
+
+      ErrorAnd<QuestionDefinition, CiviFormError> result =
+          questionService.create(questionDefinition);
+
+      if (result.isError()) {
+        String errorMessages =
+            result.getErrors().stream()
+                .map(CiviFormError::message)
+                .collect(Collectors.joining(", "));
+
+        LOGGER.error(
+            String.format(
+                "Unable to create canonical question \"%s\" due to %s",
+                questionDefinition.getName(), errorMessages));
+      } else {
+        LOGGER.info("Created canonical question \"%s\"", questionDefinition.getName());
+      }
+    }
+  }
+
+  private void inSerializableTransaction(Runnable fn, int tryCount) {
+    Transaction transaction =
+        database.beginTransaction(TxScope.requiresNew().setIsolation(TxIsolation.SERIALIZABLE));
+
+    try {
+      fn.run();
+      transaction.commit();
+    } catch (NonUniqueResultException | SerializableConflictException | RollbackException e) {
+      LOGGER.warn("Database seed transaction failed: %s", e);
+
+      if (tryCount > MAX_RETRIES) {
+        throw new RuntimeException(e);
```

Will this cause the server to crash?

I could actually see that being desired behavior. I could also see an argument for saying "let's let this fail and accept the 500s that would come from functionality that depends on it". I just think that decision needs to be made explicit in a comment.

---

Added some discussion of this to the top-level javadoc

---

Thanks for updating that!